source/extensions/formatter/cel/cel.cc

```diff
@@ -0,0 +1,97 @@
+#include "parser/parser.h"
+#endif
+
+namespace Envoy {
+namespace Extensions {
+namespace Formatter {
+
+namespace {
+
+std::string truncate(std::string str, absl::optional<size_t> max_length) {
+  if (!max_length) {
+    return str;
+  }
+
+  return str.substr(0, max_length.value());
+}
+
+} // namespace
+
+namespace Expr = Filters::Common::Expr;
+
+CELFormatter::CELFormatter(Expr::Builder& builder,
+                           const google::api::expr::v1alpha1::Expr& input_expr,
+                           absl::optional<size_t> max_length)
+    : parsed_expr_(input_expr), max_length_(max_length) {
+  compiled_expr_ = Expr::createExpression(builder, parsed_expr_);
+}
+
+absl::optional<std::string> CELFormatter::format(const Http::RequestHeaderMap& request_headers,
+                                                 const Http::ResponseHeaderMap& response_headers,
+                                                 const Http::ResponseTrailerMap& response_trailers,
+                                                 const StreamInfo::StreamInfo& stream_info,
+                                                 absl::string_view,
+                                                 AccessLog::AccessLogType) const {
+  Protobuf::Arena arena;
+  auto eval_status = Expr::evaluate(*compiled_expr_, arena, stream_info, &request_headers,
+                                    &response_headers, &response_trailers);
+  if (!eval_status.has_value() || eval_status.value().IsError()) {
+    return absl::nullopt;
+  }
+  auto result = Expr::print(eval_status.value());
+  if (max_length_.has_value()) {
+    return truncate(result, max_length_);
+  }
+  return result;
+}
+
+ProtobufWkt::Value CELFormatter::formatValue(const Http::RequestHeaderMap& request_headers,
+                                             const Http::ResponseHeaderMap& response_headers,
+                                             const Http::ResponseTrailerMap& response_trailers,
+                                             const StreamInfo::StreamInfo& stream_info,
+                                             absl::string_view, AccessLog::AccessLogType) const {
+  auto result = format(request_headers, response_headers, response_trailers, stream_info, "",
+                       AccessLog::AccessLogType::NotSet);
+  if (!result.has_value()) {
+    return ValueUtil::nullValue();
+  }
+  return ValueUtil::stringValue(result.value());
+}
+
+::Envoy::Formatter::FormatterProviderPtr
+CELFormatterCommandParser::parse(const std::string& command, const std::string& subcommand,
+                                 absl::optional<size_t>& max_length) const {
+#if defined(USE_CEL_PARSER)
+  if (command == "CEL") {
+    auto parse_status = google::api::expr::parser::Parse(subcommand);
+    if (!parse_status.ok()) {
+      throw EnvoyException("Not able to parse filter expression: " +
+                           parse_status.status().ToString());
+    }
+
+    auto expr_builder_ = Expr::createBuilder(nullptr);
```

Comment:
Suggestion: share this builder as a field in CELFormatterCommandParser:
```
BuilderPtr expr_builder_
```

Then instantiate it on the first run:

```
if (!expr_builder_) {
  expr_builder_ = Expr::createBuilder(nullptr)
}
```